// jsforce/browser/util/stream.js
import _Reflect$construct from "@babel/runtime-corejs3/core-js-stable/reflect/construct";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";
import _toConsumableArray from "@babel/runtime-corejs3/helpers/toConsumableArray";
import _classCallCheck from "@babel/runtime-corejs3/helpers/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime-corejs3/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime-corejs3/helpers/getPrototypeOf";
import _inherits from "@babel/runtime-corejs3/helpers/inherits";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
import "core-js/modules/es.error.cause.js";
import "core-js/modules/es.array.push.js";
import "core-js/modules/es.object.to-string.js";
import "core-js/modules/es.regexp.to-string.js";
import { Duplex, PassThrough, Writable } from 'stream';
export function createLazyStream() {
  var ins = new PassThrough();
  var outs = new PassThrough();
  var stream = concatStreamsAsDuplex(ins, outs);
  var piped = false;
  var setStream = function setStream(str) {
    if (piped) {
      throw new Error('stream is already piped to actual stream');
    }
    piped = true;
    ins.pipe(str).pipe(outs);
  };
  return {
    stream: stream,
    setStream: setStream
  };
}
var MemoryWriteStream = /*#__PURE__*/function (_Writable) {
  function MemoryWriteStream() {
    var _this;
    _classCallCheck(this, MemoryWriteStream);
    _this = _callSuper(this, MemoryWriteStream);
    _this._buf = Buffer.alloc(0);
    return _this;
  }
  _inherits(MemoryWriteStream, _Writable);
  return _createClass(MemoryWriteStream, [{
    key: "_write",
    value: function _write(chunk, encoding, callback) {
      this._buf = _concatInstanceProperty(Buffer).call(Buffer, [this._buf, chunk]);
      callback();
    }
  }, {
    key: "_writev",
    value: function _writev(data, callback) {
      var _context;
      this._buf = _concatInstanceProperty(Buffer).call(Buffer, _concatInstanceProperty(_context = [this._buf]).call(_context, _toConsumableArray(_mapInstanceProperty(data).call(data, function (_ref) {
        var chunk = _ref.chunk;
        return chunk;
      }))));
      callback();
    }
  }, {
    key: "toString",
    value: function toString() {
      var encoding = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'utf-8';
      return this._buf.toString(encoding);
    }
  }]);
}(Writable);
export function readAll(_x) {
  return _readAll.apply(this, arguments);
}
function _readAll() {
  _readAll = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee(rs) {
    var encoding,
      _args = arguments;
    return _regeneratorRuntime.wrap(function _callee$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          encoding = _args.length > 1 && _args[1] !== undefined ? _args[1] : 'utf-8';
          return _context2.abrupt("return", new _Promise(function (resolve, reject) {
            var ws = new MemoryWriteStream();
            rs.on('error', reject).pipe(ws).on('finish', function () {
              return resolve(ws.toString(encoding));
            });
          }));
        case 2:
        case "end":
          return _context2.stop();
      }
    }, _callee);
  }));
  return _readAll.apply(this, arguments);
}
var DuplexifiedStream = /*#__PURE__*/function (_Duplex) {
  function DuplexifiedStream(ws, rs) {
    var _opts$writableObjectM, _opts$readableObjectM;
    var _this2;
    var opts = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    _classCallCheck(this, DuplexifiedStream);
    _this2 = _callSuper(this, DuplexifiedStream, [{
      writableObjectMode: (_opts$writableObjectM = opts.writableObjectMode) !== null && _opts$writableObjectM !== void 0 ? _opts$writableObjectM : ws.writableObjectMode,
      readableObjectMode: (_opts$readableObjectM = opts.readableObjectMode) !== null && _opts$readableObjectM !== void 0 ? _opts$readableObjectM : rs.readableObjectMode
    }]);
    _this2._writable = ws;
    _this2._readable = rs;
    ws.once('finish', function () {
      _this2.end();
    });
    _this2.once('finish', function () {
      ws.end();
    });
    rs.on('readable', function () {
      _this2._readStream();
    });
    rs.once('end', function () {
      _this2.push(null);
    });
    ws.on('error', function (err) {
      return _this2.emit('error', err);
    });
    rs.on('error', function (err) {
      return _this2.emit('error', err);
    });
    return _this2;
  }
  _inherits(DuplexifiedStream, _Duplex);
  return _createClass(DuplexifiedStream, [{
    key: "_write",
    value: function _write(chunk, encoding, callback) {
      this._writable.write(chunk, encoding, callback);
    }
  }, {
    key: "_read",
    value: function _read(n) {
      this._readStream(n);
    }
  }, {
    key: "_readStream",
    value: function _readStream(n) {
      var data;
      while ((data = this._readable.read(n)) !== null) {
        this.push(data);
      }
    }
  }]);
}(Duplex);
export function concatStreamsAsDuplex(ws, rs, opts) {
  return new DuplexifiedStream(ws, rs, opts);
}
