// jsforce/browser/browser/canvas.js
import "core-js/modules/es.error.cause.js";
import "core-js/modules/es.function.name.js";
import "core-js/modules/es.regexp.test.js";
import _sliceInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/slice";
import _Array$from from "@babel/runtime-corejs3/core-js-stable/array/from";
import _Symbol from "@babel/runtime-corejs3/core-js-stable/symbol";
import _getIteratorMethod from "@babel/runtime-corejs3/core-js/get-iterator-method";
import _Array$isArray from "@babel/runtime-corejs3/core-js-stable/array/is-array";
import _typeof from "@babel/runtime-corejs3/helpers/typeof";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime-corejs3/helpers/slicedToArray";
import _Object$keys2 from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
import "core-js/modules/es.array.join.js";
import "core-js/modules/es.array.push.js";
import "core-js/modules/es.object.to-string.js";
import "core-js/modules/es.regexp.exec.js";
import "core-js/modules/es.regexp.to-string.js";
import "core-js/modules/es.string.split.js";
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof _Symbol && _getIteratorMethod(r) || r["@@iterator"]; if (!t) { if (_Array$isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { var _context3; if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = _sliceInstanceProperty(_context3 = {}.toString.call(r)).call(_context3, 8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? _Array$from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
/**
 *
 */
import { Transform } from 'stream';
function parseHeaders(hs) {
  var headers = {};
  var _iterator = _createForOfIteratorHelper(hs.split(/\n/)),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var line = _step.value;
      var _line$split = line.split(/\s*:\s*/),
        _line$split2 = _slicedToArray(_line$split, 2),
        name = _line$split2[0],
        value = _line$split2[1];
      headers[name.toLowerCase()] = value;
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  return headers;
}
function processCanvasRequest(_x, _x2, _x3) {
  return _processCanvasRequest.apply(this, arguments);
}
function _processCanvasRequest() {
  _processCanvasRequest = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee2(params, signedRequest, requestBody) {
    var settings, paramHeaders, _i, _Object$keys, name, data, headers, responseBody;
    return _regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          settings = {
            client: signedRequest.client,
            method: params.method,
            data: requestBody
          };
          paramHeaders = params.headers;
          if (paramHeaders) {
            settings.headers = {};
            for (_i = 0, _Object$keys = _Object$keys2(paramHeaders); _i < _Object$keys.length; _i++) {
              name = _Object$keys[_i];
              if (name.toLowerCase() === 'content-type') {
                settings.contentType = paramHeaders[name];
              } else {
                settings.headers[name] = paramHeaders[name];
              }
            }
          }
          _context2.next = 5;
          return new _Promise(function (resolve, reject) {
            settings.success = resolve;
            settings.failure = reject;
            Sfdc.canvas.client.ajax(params.url, settings);
          });
        case 5:
          data = _context2.sent;
          headers = parseHeaders(data.responseHeaders);
          responseBody = data.payload;
          if (typeof responseBody !== 'string') {
            responseBody = _JSON$stringify(responseBody);
          }
          return _context2.abrupt("return", {
            statusCode: data.status,
            headers: headers,
            body: responseBody
          });
        case 10:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return _processCanvasRequest.apply(this, arguments);
}
function createRequest(signedRequest) {
  return function (params) {
    var buf = [];
    var stream = new Transform({
      transform: function transform(chunk, encoding, callback) {
        buf.push(typeof chunk === 'string' ? chunk : chunk.toString('utf8'));
        callback();
      },
      flush: function flush() {
        _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
          var body, response;
          return _regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) switch (_context.prev = _context.next) {
              case 0:
                body = buf.join('');
                _context.next = 3;
                return processCanvasRequest(params, signedRequest, body);
              case 3:
                response = _context.sent;
                stream.emit('response', response);
                stream.emit('complete', response);
                stream.push(response.body);
                stream.push(null);
              case 8:
              case "end":
                return _context.stop();
            }
          }, _callee);
        }))();
      }
    });
    if (params.body) {
      stream.end(params.body);
    }
    return stream;
  };
}
export default {
  supported: (typeof Sfdc === "undefined" ? "undefined" : _typeof(Sfdc)) === 'object' && typeof Sfdc.canvas !== 'undefined',
  createRequest: createRequest
};
