// jsforce/browser/browser/client.js
import _Reflect$construct from "@babel/runtime-corejs3/core-js-stable/reflect/construct";
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _filterInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/filter";
import _Object$getOwnPropertyDescriptor from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _Object$getOwnPropertyDescriptors from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors";
import _Object$defineProperties from "@babel/runtime-corejs3/core-js-stable/object/define-properties";
import _Object$defineProperty from "@babel/runtime-corejs3/core-js-stable/object/define-property";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";
import _classCallCheck from "@babel/runtime-corejs3/helpers/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime-corejs3/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime-corejs3/helpers/getPrototypeOf";
import _inherits from "@babel/runtime-corejs3/helpers/inherits";
import _slicedToArray from "@babel/runtime-corejs3/helpers/slicedToArray";
function ownKeys(e, r) { var t = _Object$keys(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = _filterInstanceProperty(o).call(o, function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context5, _context6; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty(_context5 = ownKeys(Object(t), !0)).call(_context5, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : _forEachInstanceProperty(_context6 = ownKeys(Object(t))).call(_context6, function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
import "core-js/modules/es.error.cause.js";
import "core-js/modules/es.array.join.js";
import "core-js/modules/es.number.constructor.js";
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.array.push.js";
import "core-js/modules/es.object.to-string.js";
import "core-js/modules/es.regexp.constructor.js";
import "core-js/modules/es.regexp.dot-all.js";
import "core-js/modules/es.regexp.exec.js";
import "core-js/modules/es.regexp.sticky.js";
import "core-js/modules/es.regexp.to-string.js";
import "core-js/modules/es.string.match.js";
import "core-js/modules/es.string.search.js";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _Object$fromEntries from "@babel/runtime-corejs3/core-js-stable/object/from-entries";
import _URLSearchParams from "@babel/runtime-corejs3/core-js-stable/url-search-params";
import _setTimeout from "@babel/runtime-corejs3/core-js-stable/set-timeout";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import _setInterval from "@babel/runtime-corejs3/core-js-stable/set-interval";
import _Date$now from "@babel/runtime-corejs3/core-js-stable/date/now";
import _reverseInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/reverse";
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
/**
 * @file Browser client connection management class
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */
import { EventEmitter } from 'events';
import Connection from '../connection';
import OAuth2 from '../oauth2';

/**
 * @private
 */
function popupWin(url, w, h) {
  var _context, _context2, _context3;
  var left = screen.width / 2 - w / 2;
  var top = screen.height / 2 - h / 2;
  return window.open(url, undefined, _concatInstanceProperty(_context = _concatInstanceProperty(_context2 = _concatInstanceProperty(_context3 = "location=yes,toolbar=no,status=no,menubar=no,width=".concat(w, ",height=")).call(_context3, h, ",top=")).call(_context2, top, ",left=")).call(_context, left));
}

/**
 * @private
 */
function handleCallbackResponse() {
  var res = checkCallbackResponse();
  var state = localStorage.getItem('jsforce_state');
  if (res && state && res.body.get('state') === state) {
    localStorage.removeItem('jsforce_state');
    var _state$split = state.split('.'),
      _state$split2 = _slicedToArray(_state$split, 2),
      prefix = _state$split2[0],
      promptType = _state$split2[1];
    var cli = new BrowserClient(prefix);
    if (res.success) {
      cli._storeTokens(_Object$fromEntries(res.body));
      location.hash = '';
    } else {
      cli._storeError(res.body);
    }
    if (promptType === 'popup') {
      window.close();
    }
    return true;
  }
}

/**
 * @private
 */
function checkCallbackResponse() {
  var params;
  if (window.location.hash) {
    params = new _URLSearchParams(window.location.hash.substring(1));
    if (params.get('access_token')) {
      return {
        success: true,
        body: params
      };
    }
  } else if (window.location.search) {
    params = new _URLSearchParams(window.location.search.substring(1));
    if (params.get('error')) {
      return {
        success: false,
        body: params
      };
    }
  }
}

/**
 *
 */

/**
 *
 */
var DEFAULT_POPUP_WIN_WIDTH = 912;
var DEFAULT_POPUP_WIN_HEIGHT = 513;

/** @private **/
var clientIdx = 0;

/**
 *
 */
export var BrowserClient = /*#__PURE__*/function (_EventEmitter) {
  /**
   *
   */
  function BrowserClient(prefix) {
    var _this;
    _classCallCheck(this, BrowserClient);
    _this = _callSuper(this, BrowserClient);
    _this._prefix = prefix || 'jsforce' + clientIdx++;
    return _this;
  }
  _inherits(BrowserClient, _EventEmitter);
  return _createClass(BrowserClient, [{
    key: "connection",
    get: function get() {
      if (!this._connection) {
        this._connection = new Connection(this._config);
      }
      return this._connection;
    }

    /**
     *
     */
  }, {
    key: "init",
    value: function init(config) {
      var _this2 = this;
      if (handleCallbackResponse()) {
        return;
      }
      this._config = config;
      var tokens = this._getTokens();
      if (tokens) {
        this.connection._establish(tokens);
        _setTimeout(function () {
          _this2.emit('connect', _this2.connection);
        }, 10);
      }
    }

    /**
     *
     */
  }, {
    key: "login",
    value: function login() {
      var _this$_config,
        _size$width,
        _size$height,
        _this3 = this;
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var scope = options.scope,
        size = options.size;
      var oauth2 = new OAuth2((_this$_config = this._config) !== null && _this$_config !== void 0 ? _this$_config : {});
      var rand = Math.random().toString(36).substring(2);
      var state = [this._prefix, 'popup', rand].join('.');
      localStorage.setItem('jsforce_state', state);
      var authzUrl = oauth2.getAuthorizationUrl(_objectSpread({
        response_type: 'token',
        state: state
      }, scope ? {
        scope: scope
      } : {}));
      var pw = popupWin(authzUrl, (_size$width = size === null || size === void 0 ? void 0 : size.width) !== null && _size$width !== void 0 ? _size$width : DEFAULT_POPUP_WIN_WIDTH, (_size$height = size === null || size === void 0 ? void 0 : size.height) !== null && _size$height !== void 0 ? _size$height : DEFAULT_POPUP_WIN_HEIGHT);
      return new _Promise(function (resolve, reject) {
        if (!pw) {
          var _state = [_this3._prefix, 'redirect', rand].join('.');
          localStorage.setItem('jsforce_state', _state);
          var _authzUrl = oauth2.getAuthorizationUrl(_objectSpread({
            response_type: 'token',
            state: _state
          }, scope ? {
            scope: scope
          } : {}));
          location.href = _authzUrl;
          return;
        }
        _this3._removeTokens();
        var pid = _setInterval(function () {
          try {
            if (!pw || pw.closed) {
              clearInterval(pid);
              var tokens = _this3._getTokens();
              if (tokens) {
                _this3.connection._establish(tokens);
                _this3.emit('connect', _this3.connection);
                resolve({
                  status: 'connect'
                });
              } else {
                var err = _this3._getError();
                if (err) {
                  reject(new Error(err.error + ': ' + err.error_description));
                } else {
                  resolve({
                    status: 'cancel'
                  });
                }
              }
            }
          } catch (e) {
            //
          }
        }, 1000);
      });
    }

    /**
     *
     */
  }, {
    key: "isLoggedIn",
    value: function isLoggedIn() {
      return !!this.connection.accessToken;
    }

    /**
     *
     */
  }, {
    key: "logout",
    value: function logout() {
      this.connection.logout();
      this._removeTokens();
      this.emit('disconnect');
    }

    /**
     * @private
     */
  }, {
    key: "_getTokens",
    value: function _getTokens() {
      var regexp = new RegExp('(^|;\\s*)' + this._prefix + '_loggedin=true(;|$)');
      if (document.cookie.match(regexp)) {
        var issuedAt = Number(localStorage.getItem(this._prefix + '_issued_at'));
        // 2 hours
        if (_Date$now() < issuedAt + 2 * 60 * 60 * 1000) {
          var userInfo;
          var idUrl = localStorage.getItem(this._prefix + '_id');
          if (idUrl) {
            var _context4;
            var _idUrl$split$reverse = _reverseInstanceProperty(_context4 = idUrl.split('/')).call(_context4),
              _idUrl$split$reverse2 = _slicedToArray(_idUrl$split$reverse, 2),
              id = _idUrl$split$reverse2[0],
              organizationId = _idUrl$split$reverse2[1];
            userInfo = {
              id: id,
              organizationId: organizationId,
              url: idUrl
            };
          }
          return {
            accessToken: localStorage.getItem(this._prefix + '_access_token'),
            instanceUrl: localStorage.getItem(this._prefix + '_instance_url'),
            userInfo: userInfo
          };
        }
      }
      return null;
    }

    /**
     * @private
     */
  }, {
    key: "_storeTokens",
    value: function _storeTokens(params) {
      localStorage.setItem(this._prefix + '_access_token', params.access_token);
      localStorage.setItem(this._prefix + '_instance_url', params.instance_url);
      localStorage.setItem(this._prefix + '_issued_at', params.issued_at);
      localStorage.setItem(this._prefix + '_id', params.id);
      document.cookie = this._prefix + '_loggedin=true;';
    }

    /**
     * @private
     */
  }, {
    key: "_removeTokens",
    value: function _removeTokens() {
      localStorage.removeItem(this._prefix + '_access_token');
      localStorage.removeItem(this._prefix + '_instance_url');
      localStorage.removeItem(this._prefix + '_issued_at');
      localStorage.removeItem(this._prefix + '_id');
      document.cookie = this._prefix + '_loggedin=';
    }

    /**
     * @private
     */
  }, {
    key: "_getError",
    value: function _getError() {
      try {
        var _localStorage$getItem;
        var err = JSON.parse((_localStorage$getItem = localStorage.getItem(this._prefix + '_error')) !== null && _localStorage$getItem !== void 0 ? _localStorage$getItem : '');
        localStorage.removeItem(this._prefix + '_error');
        return err;
      } catch (e) {
        //
      }
    }

    /**
     * @private
     */
  }, {
    key: "_storeError",
    value: function _storeError(err) {
      localStorage.setItem(this._prefix + '_error', _JSON$stringify(err));
    }
  }]);
}(EventEmitter);

/**
 *
 */
var client = new BrowserClient();
export default client;
