// jsforce/browser/request-helper.js
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _filterInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/filter";
import _Object$getOwnPropertyDescriptor from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _Object$getOwnPropertyDescriptors from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors";
import _Object$defineProperties from "@babel/runtime-corejs3/core-js-stable/object/define-properties";
import _Object$defineProperty from "@babel/runtime-corejs3/core-js-stable/object/define-property";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
function ownKeys(e, r) { var t = _Object$keys(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = _filterInstanceProperty(o).call(o, function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context3, _context4; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty(_context3 = ownKeys(Object(t), !0)).call(_context3, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : _forEachInstanceProperty(_context4 = ownKeys(Object(t))).call(_context4, function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
import "core-js/modules/es.error.cause.js";
import "core-js/modules/es.array.push.js";
import "core-js/modules/es.number.to-fixed.js";
import _setTimeout from "@babel/runtime-corejs3/core-js-stable/set-timeout";
import _Date$now from "@babel/runtime-corejs3/core-js-stable/date/now";
import _Set from "@babel/runtime-corejs3/core-js-stable/set";
import { PassThrough } from 'stream';
import { concatStreamsAsDuplex, readAll } from './util/stream';
import FormData from 'form-data';
import { getLogger } from './util/logger';
var logger = getLogger('request-helper');

/**
 *
 */
export function createHttpRequestHandlerStreams(req) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var reqBody = req.body;
  var input = new PassThrough();
  var output = new PassThrough();
  var duplex = concatStreamsAsDuplex(input, output);
  if (typeof reqBody !== 'undefined') {
    _setTimeout(function () {
      if (reqBody instanceof FormData) {
        duplex.end(reqBody.getBuffer());
      } else {
        duplex.end(reqBody, 'utf8');
      }
    }, 0);
  }
  duplex.on('response', /*#__PURE__*/function () {
    var _ref = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee(res) {
      var start, resBody;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            if (!(duplex.listenerCount('complete') > 0)) {
              _context.next = 8;
              break;
            }
            logger.debug('Processing response');
            start = _Date$now();
            _context.next = 5;
            return readAll(duplex, options.encoding);
          case 5:
            resBody = _context.sent;
            logger.debug("Response body read: ".concat((resBody.length / 1024 / 1024).toFixed(1), "MB ") + "in ".concat(_Date$now() - start, "ms"));
            duplex.emit('complete', _objectSpread(_objectSpread({}, res), {}, {
              body: resBody
            }));
          case 8:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function (_x) {
      return _ref.apply(this, arguments);
    };
  }());
  return {
    input: input,
    output: output,
    stream: duplex
  };
}
var redirectStatuses = new _Set([301, 302, 303, 307, 308]);

/**
 *
 */
export function isRedirect(status) {
  return redirectStatuses.has(status);
}

/**
 *
 */
var MAX_REDIRECT_COUNT = 10;

/**
 *
 */
export function performRedirectRequest(req, res, followRedirect, counter, redirectCallback) {
  if (counter >= MAX_REDIRECT_COUNT) {
    throw new Error('Reached to maximum redirect count');
  }
  var redirectUrl = res.headers['location'];
  if (!redirectUrl) {
    throw new Error('No redirect URI found');
  }
  var getRedirectRequest = typeof followRedirect === 'function' ? followRedirect : function () {
    return {
      method: 'GET',
      url: redirectUrl,
      headers: req.headers
    };
  };
  var nextReqParams = getRedirectRequest(redirectUrl);
  if (!nextReqParams) {
    throw new Error('Cannot handle redirect for ' + redirectUrl);
  }
  redirectCallback(nextReqParams);
}

/**
 *
 */
export function executeWithTimeout(_x2, _x3, _x4) {
  return _executeWithTimeout.apply(this, arguments);
}
function _executeWithTimeout() {
  _executeWithTimeout = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee2(execFn, msec, cancelCallback) {
    var timeout, pid, res;
    return _regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          timeout = false;
          pid = msec != null ? _setTimeout(function () {
            timeout = true;
            cancelCallback === null || cancelCallback === void 0 || cancelCallback();
          }, msec) : undefined;
          _context2.prev = 2;
          _context2.next = 5;
          return execFn();
        case 5:
          res = _context2.sent;
        case 6:
          _context2.prev = 6;
          if (pid) {
            clearTimeout(pid);
          }
          return _context2.finish(6);
        case 9:
          if (!timeout) {
            _context2.next = 11;
            break;
          }
          throw new Error('Request Timeout');
        case 11:
          return _context2.abrupt("return", res);
        case 12:
        case "end":
          return _context2.stop();
      }
    }, _callee2, null, [[2,, 6, 9]]);
  }));
  return _executeWithTimeout.apply(this, arguments);
}
