// jsforce/browser/api/analytics.js
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _filterInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/filter";
import _Object$getOwnPropertyDescriptor from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _Object$getOwnPropertyDescriptors from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors";
import _Object$defineProperties from "@babel/runtime-corejs3/core-js-stable/object/define-properties";
import _Object$defineProperty from "@babel/runtime-corejs3/core-js-stable/object/define-property";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";
import _classCallCheck from "@babel/runtime-corejs3/helpers/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/createClass";
function ownKeys(e, r) { var t = _Object$keys(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = _filterInstanceProperty(o).call(o, function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context, _context2; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty(_context = ownKeys(Object(t), !0)).call(_context, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : _forEachInstanceProperty(_context2 = ownKeys(Object(t))).call(_context2, function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _Array$isArray from "@babel/runtime-corejs3/core-js-stable/array/is-array";
import "core-js/modules/es.array.join.js";
import "core-js/modules/es.array.push.js";
/**
 * @file Manages Salesforce Analytics API
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */
import { registerModule } from '../jsforce';

/*----------------------------------------------------------------------------------*/

/*----------------------------------------------------------------------------------*/
/**
 * Report object class in Analytics API
 */
export var ReportInstance = /*#__PURE__*/function () {
  /**
   *
   */
  function ReportInstance(report, id) {
    _classCallCheck(this, ReportInstance);
    this._report = report;
    this._conn = report._conn;
    this.id = id;
  }

  /**
   * Retrieve report result asynchronously executed
   */
  return _createClass(ReportInstance, [{
    key: "retrieve",
    value: function retrieve() {
      var url = [this._conn._baseUrl(), 'analytics', 'reports', this._report.id, 'instances', this.id].join('/');
      return this._conn.request(url);
    }
  }]);
}();

/*----------------------------------------------------------------------------------*/
/**
 * Report object class in Analytics API
 */
export var Report = /*#__PURE__*/function () {
  /**
   *
   */
  function Report(conn, id) {
    _classCallCheck(this, Report);
    /**
     * Synonym of Analytics~Report#destroy()
     */
    _defineProperty(this, "delete", this.destroy);
    /**
     * Synonym of Analytics~Report#destroy()
     */
    _defineProperty(this, "del", this.destroy);
    /**
     * Synonym of Analytics~Report#execute()
     */
    _defineProperty(this, "run", this.execute);
    /**
     * Synonym of Analytics~Report#execute()
     */
    _defineProperty(this, "exec", this.execute);
    this._conn = conn;
    this.id = id;
  }

  /**
   * Describe report metadata
   */
  return _createClass(Report, [{
    key: "describe",
    value: function describe() {
      var url = [this._conn._baseUrl(), 'analytics', 'reports', this.id, 'describe'].join('/');
      return this._conn.request(url);
    }

    /**
     * Destroy a report
     */
  }, {
    key: "destroy",
    value: function destroy() {
      var url = [this._conn._baseUrl(), 'analytics', 'reports', this.id].join('/');
      return this._conn.request({
        method: 'DELETE',
        url: url
      });
    }
  }, {
    key: "clone",
    value:
    /**
     * Clones a given report
     */
    function clone(name) {
      var url = [this._conn._baseUrl(), 'analytics', 'reports'].join('/') + '?cloneId=' + this.id;
      var config = {
        reportMetadata: {
          name: name
        }
      };
      return this._conn.request({
        method: 'POST',
        url: url,
        headers: {
          'Content-Type': 'application/json'
        },
        body: _JSON$stringify(config)
      });
    }

    /**
     * Explain plan for executing report
     */
  }, {
    key: "explain",
    value: function explain() {
      var url = '/query/?explain=' + this.id;
      return this._conn.request(url);
    }

    /**
     * Run report synchronously
     */
  }, {
    key: "execute",
    value: function execute() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var url = [this._conn._baseUrl(), 'analytics', 'reports', this.id].join('/') + '?includeDetails=' + (options.details ? 'true' : 'false');
      return this._conn.request(_objectSpread({
        url: url
      }, options.metadata ? {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: _JSON$stringify(options.metadata)
      } : {
        method: 'GET'
      }));
    }
  }, {
    key: "executeAsync",
    value:
    /**
     * Run report asynchronously
     */
    function executeAsync() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var url = [this._conn._baseUrl(), 'analytics', 'reports', this.id, 'instances'].join('/') + (options.details ? '?includeDetails=true' : '');
      return this._conn.request(_objectSpread({
        method: 'POST',
        url: url
      }, options.metadata ? {
        headers: {
          'Content-Type': 'application/json'
        },
        body: _JSON$stringify(options.metadata)
      } : {
        body: ''
      }));
    }

    /**
     * Get report instance for specified instance ID
     */
  }, {
    key: "instance",
    value: function instance(id) {
      return new ReportInstance(this, id);
    }

    /**
     * List report instances which had been executed asynchronously
     */
  }, {
    key: "instances",
    value: function instances() {
      var url = [this._conn._baseUrl(), 'analytics', 'reports', this.id, 'instances'].join('/');
      return this._conn.request(url);
    }
  }]);
}();

/*----------------------------------------------------------------------------------*/
/**
 * Dashboard object class in the Analytics API
 */
export var Dashboard = /*#__PURE__*/function () {
  /**
   *
   */
  function Dashboard(conn, id) {
    _classCallCheck(this, Dashboard);
    /**
     * Synonym of Analytics~Dashboard#destroy()
     */
    _defineProperty(this, "delete", this.destroy);
    /**
     * Synonym of Analytics~Dashboard#destroy()
     */
    _defineProperty(this, "del", this.destroy);
    this._conn = conn;
    this.id = id;
  }

  /**
   * Describe dashboard metadata
   *
   * @method Analytics~Dashboard#describe
   * @param {Callback.<Analytics-DashboardMetadata>} [callback] - Callback function
   * @returns {Promise.<Analytics-DashboardMetadata>}
   */
  return _createClass(Dashboard, [{
    key: "describe",
    value: function describe() {
      var url = [this._conn._baseUrl(), 'analytics', 'dashboards', this.id, 'describe'].join('/');
      return this._conn.request(url);
    }

    /**
     * Get details about dashboard components
     */
  }, {
    key: "components",
    value: function components(componentIds) {
      var url = [this._conn._baseUrl(), 'analytics', 'dashboards', this.id].join('/');
      var config = {
        componentIds: _Array$isArray(componentIds) ? componentIds : typeof componentIds === 'string' ? [componentIds] : undefined
      };
      return this._conn.request({
        method: 'POST',
        url: url,
        headers: {
          'Content-Type': 'application/json'
        },
        body: _JSON$stringify(config)
      });
    }

    /**
     * Get dashboard status
     */
  }, {
    key: "status",
    value: function status() {
      var url = [this._conn._baseUrl(), 'analytics', 'dashboards', this.id, 'status'].join('/');
      return this._conn.request(url);
    }

    /**
     * Refresh a dashboard
     */
  }, {
    key: "refresh",
    value: function refresh() {
      var url = [this._conn._baseUrl(), 'analytics', 'dashboards', this.id].join('/');
      return this._conn.request({
        method: 'PUT',
        url: url,
        body: ''
      });
    }

    /**
     * Clone a dashboard
     */
  }, {
    key: "clone",
    value: function clone(config, folderId) {
      var url = [this._conn._baseUrl(), 'analytics', 'dashboards'].join('/') + '?cloneId=' + this.id;
      if (typeof config === 'string') {
        config = {
          name: config,
          folderId: folderId
        };
      }
      return this._conn.request({
        method: 'POST',
        url: url,
        headers: {
          'Content-Type': 'application/json'
        },
        body: _JSON$stringify(config)
      });
    }

    /**
     * Destroy a dashboard
     */
  }, {
    key: "destroy",
    value: function destroy() {
      var url = [this._conn._baseUrl(), 'analytics', 'dashboards', this.id].join('/');
      return this._conn.request({
        method: 'DELETE',
        url: url
      });
    }
  }]);
}();

/*----------------------------------------------------------------------------------*/
/**
 * API class for Analytics API
 */
export var Analytics = /*#__PURE__*/function () {
  /**
   *
   */
  function Analytics(conn) {
    _classCallCheck(this, Analytics);
    this._conn = conn;
  }

  /**
   * Get report object of Analytics API
   */
  return _createClass(Analytics, [{
    key: "report",
    value: function report(id) {
      return new Report(this._conn, id);
    }

    /**
     * Get recent report list
     */
  }, {
    key: "reports",
    value: function reports() {
      var url = [this._conn._baseUrl(), 'analytics', 'reports'].join('/');
      return this._conn.request(url);
    }

    /**
     * Get dashboard object of Analytics API
     */
  }, {
    key: "dashboard",
    value: function dashboard(id) {
      return new Dashboard(this._conn, id);
    }

    /**
     * Get recent dashboard list
     */
  }, {
    key: "dashboards",
    value: function dashboards() {
      var url = [this._conn._baseUrl(), 'analytics', 'dashboards'].join('/');
      return this._conn.request(url);
    }
  }]);
}();

/*--------------------------------------------*/
/*
 * Register hook in connection instantiation for dynamically adding this API module features
 */
registerModule('analytics', function (conn) {
  return new Analytics(conn);
});
export default Analytics;
