// jsforce/browser/api/chatter.js
import "core-js/modules/es.array.push.js";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _filterInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/filter";
import _Object$getOwnPropertyDescriptor from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor";
import _Object$getOwnPropertyDescriptors from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors";
import _Object$defineProperties from "@babel/runtime-corejs3/core-js-stable/object/define-properties";
import _Object$defineProperty from "@babel/runtime-corejs3/core-js-stable/object/define-property";
import _Reflect$construct from "@babel/runtime-corejs3/core-js-stable/reflect/construct";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";
import _possibleConstructorReturn from "@babel/runtime-corejs3/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime-corejs3/helpers/getPrototypeOf";
import _inherits from "@babel/runtime-corejs3/helpers/inherits";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";
import _classCallCheck from "@babel/runtime-corejs3/helpers/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/createClass";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _indexOfInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/index-of";
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _startsWithInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/starts-with";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import "core-js/modules/es.array.join.js";
import "core-js/modules/es.regexp.exec.js";
import "core-js/modules/es.regexp.test.js";
function ownKeys(e, r) { var t = _Object$keys(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = _filterInstanceProperty(o).call(o, function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context5, _context6; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty(_context5 = ownKeys(Object(t), !0)).call(_context5, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : _forEachInstanceProperty(_context6 = ownKeys(Object(t))).call(_context6, function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * @file Manages Salesforce Chatter REST API calls
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */
import { registerModule } from '../jsforce';
import { isObject } from '../util/function';

/**
 *
 */
/*--------------------------------------------*/
/**
 * A class representing chatter API request
 */
var Request = /*#__PURE__*/function () {
  function Request(chatter, request) {
    _classCallCheck(this, Request);
    this._chatter = chatter;
    this._request = request;
  }

  /**
   * Retrieve parameters in batch request form
   */
  return _createClass(Request, [{
    key: "batchParams",
    value: function batchParams() {
      var _this$_request = this._request,
        method = _this$_request.method,
        url = _this$_request.url,
        body = _this$_request.body;
      return _objectSpread({
        method: method,
        url: this._chatter._normalizeUrl(url)
      }, typeof body !== 'undefined' ? {
        richInput: body
      } : {});
    }

    /**
     * Retrieve parameters in batch request form
     *
     * @method Chatter~Request#promise
     * @returns {Promise.<Chatter~RequestResult>}
     */
  }, {
    key: "promise",
    value: function promise() {
      return this._promise || (this._promise = this._chatter._request(this._request));
    }

    /**
     * Returns Node.js Stream object for request
     *
     * @method Chatter~Request#stream
     * @returns {stream.Stream}
     */
  }, {
    key: "stream",
    value: function stream() {
      return this._chatter._request(this._request).stream();
    }

    /**
     * Promise/A+ interface
     * http://promises-aplus.github.io/promises-spec/
     *
     * Delegate to deferred promise, return promise instance for batch result
     */
  }, {
    key: "then",
    value: function then(onResolve, onReject) {
      return this.promise().then(onResolve, onReject);
    }
  }]);
}();
function apppendQueryParamsToUrl(url, queryParams) {
  if (queryParams) {
    var _context;
    var qstring = _mapInstanceProperty(_context = _Object$keys(queryParams)).call(_context, function (name) {
      var _context2, _queryParams$name;
      return _concatInstanceProperty(_context2 = "".concat(name, "=")).call(_context2, encodeURIComponent(String((_queryParams$name = queryParams[name]) !== null && _queryParams$name !== void 0 ? _queryParams$name : '')));
    }).join('&');
    url += (_indexOfInstanceProperty(url).call(url, '?') > 0 ? '&' : '?') + qstring;
  }
  return url;
}

/*------------------------------*/
export var Resource = /*#__PURE__*/function (_Request) {
  /**
   *
   */
  function Resource(chatter, url, queryParams) {
    var _this;
    _classCallCheck(this, Resource);
    _this = _callSuper(this, Resource, [chatter, {
      method: 'GET',
      url: apppendQueryParamsToUrl(url, queryParams)
    }]);
    /**
     * Synonym of Resource#destroy()
     */
    _defineProperty(_this, "delete", _this.destroy);
    /**
     * Synonym of Resource#destroy()
     */
    _defineProperty(_this, "del", _this.destroy);
    _this._url = _this._request.url;
    return _this;
  }

  /**
   * Create a new resource
   */
  _inherits(Resource, _Request);
  return _createClass(Resource, [{
    key: "create",
    value: function create(data) {
      return this._chatter.request({
        method: 'POST',
        url: this._url,
        body: data
      });
    }

    /**
     * Retrieve resource content
     */
  }, {
    key: "retrieve",
    value: function retrieve() {
      return this._chatter.request({
        method: 'GET',
        url: this._url
      });
    }

    /**
     * Update specified resource
     */
  }, {
    key: "update",
    value: function update(data) {
      return this._chatter.request({
        method: 'POST',
        url: this._url,
        body: data
      });
    }

    /**
     * Delete specified resource
     */
  }, {
    key: "destroy",
    value: function destroy() {
      return this._chatter.request({
        method: 'DELETE',
        url: this._url
      });
    }
  }]);
}(Request);

/*------------------------------*/
/**
 * API class for Chatter REST API call
 */
export var Chatter = /*#__PURE__*/function () {
  /**
   *
   */
  function Chatter(conn) {
    _classCallCheck(this, Chatter);
    this._conn = conn;
  }

  /**
   * Sending request to API endpoint
   * @private
   */
  return _createClass(Chatter, [{
    key: "_request",
    value: function _request(req_) {
      var method = req_.method,
        url_ = req_.url,
        headers_ = req_.headers,
        body_ = req_.body;
      var headers = headers_ !== null && headers_ !== void 0 ? headers_ : {};
      var body;
      if (/^(put|post|patch)$/i.test(method)) {
        if (isObject(body_)) {
          headers = _objectSpread(_objectSpread({}, headers_), {}, {
            'Content-Type': 'application/json'
          });
          body = _JSON$stringify(body_);
        } else {
          body = body_;
        }
      }
      var url = this._normalizeUrl(url_);
      return this._conn.request({
        method: method,
        url: url,
        headers: headers,
        body: body
      });
    }

    /**
     * Convert path to site root relative url
     * @private
     */
  }, {
    key: "_normalizeUrl",
    value: function _normalizeUrl(url) {
      if (_startsWithInstanceProperty(url).call(url, '/chatter/') || _startsWithInstanceProperty(url).call(url, '/connect/')) {
        return '/services/data/v' + this._conn.version + url;
      } else if (/^\/v[\d]+\.[\d]+\//.test(url)) {
        return '/services/data' + url;
      } else if (!_startsWithInstanceProperty(url).call(url, '/services/') && _startsWithInstanceProperty(url).call(url, '/')) {
        return '/services/data/v' + this._conn.version + '/chatter' + url;
      } else {
        return url;
      }
    }

    /**
     * Make a request for chatter API resource
     */
  }, {
    key: "request",
    value: function request(req) {
      return new Request(this, req);
    }

    /**
     * Make a resource request to chatter API
     */
  }, {
    key: "resource",
    value: function resource(url, queryParams) {
      return new Resource(this, url, queryParams);
    }

    /**
     * Make a batch request to chatter API
     */
  }, {
    key: "batch",
    value: (function () {
      var _batch = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime.mark(function _callee(requests) {
        var _context3;
        var deferreds, res;
        return _regeneratorRuntime.wrap(function _callee$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              deferreds = _mapInstanceProperty(requests).call(requests, function (request) {
                var deferred = defer();
                request._promise = deferred.promise;
                return deferred;
              });
              _context4.next = 3;
              return this.request({
                method: 'POST',
                url: this._normalizeUrl('/connect/batch'),
                body: {
                  batchRequests: _mapInstanceProperty(requests).call(requests, function (request) {
                    return request.batchParams();
                  })
                }
              });
            case 3:
              res = _context4.sent;
              _forEachInstanceProperty(_context3 = res.results).call(_context3, function (result, i) {
                var deferred = deferreds[i];
                if (result.statusCode >= 400) {
                  deferred.reject(result.result);
                } else {
                  deferred.resolve(result.result);
                }
              });
              return _context4.abrupt("return", res);
            case 6:
            case "end":
              return _context4.stop();
          }
        }, _callee, this);
      }));
      function batch(_x) {
        return _batch.apply(this, arguments);
      }
      return batch;
    }())
  }]);
}();
function defer() {
  var resolve_ = function resolve_() {};
  var reject_ = function reject_() {};
  var promise = new _Promise(function (resolve, reject) {
    resolve_ = resolve;
    reject_ = reject;
  });
  return {
    promise: promise,
    resolve: resolve_,
    reject: reject_
  };
}

/*--------------------------------------------*/
/*
 * Register hook in connection instantiation for dynamically adding this API module features
 */
registerModule('chatter', function (conn) {
  return new Chatter(conn);
});
export default Chatter;
