// jsforce/browser/api/tooling.js
import "core-js/modules/es.error.cause.js";
import "core-js/modules/es.object.to-string.js";
import "core-js/modules/es.regexp.exec.js";
import "core-js/modules/es.regexp.test.js";
import "core-js/modules/es.regexp.to-string.js";
import _sliceInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/slice";
import _Array$from from "@babel/runtime-corejs3/core-js-stable/array/from";
import _Symbol from "@babel/runtime-corejs3/core-js-stable/symbol";
import _getIteratorMethod from "@babel/runtime-corejs3/core-js/get-iterator-method";
import _Array$isArray from "@babel/runtime-corejs3/core-js-stable/array/is-array";
import _classCallCheck from "@babel/runtime-corejs3/helpers/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/createClass";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";
import "core-js/modules/es.function.name.js";
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof _Symbol && _getIteratorMethod(r) || r["@@iterator"]; if (!t) { if (_Array$isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { var _context; if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = _sliceInstanceProperty(_context = {}.toString.call(r)).call(_context, 8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? _Array$from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
/**
 * @file Manages Tooling APIs
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */
import { registerModule } from '../jsforce';
import Connection from '../connection';
import Cache from '../cache';

/**
 *
 */

/**
 *
 */
var _Connection$prototype = Connection.prototype,
  query = _Connection$prototype.query,
  queryMore = _Connection$prototype.queryMore,
  _ensureVersion = _Connection$prototype._ensureVersion,
  create = _Connection$prototype.create,
  _createSingle = _Connection$prototype._createSingle,
  _createMany = _Connection$prototype._createMany,
  _createParallel = _Connection$prototype._createParallel,
  retrieve = _Connection$prototype.retrieve,
  _retrieveSingle = _Connection$prototype._retrieveSingle,
  _retrieveParallel = _Connection$prototype._retrieveParallel,
  _retrieveMany = _Connection$prototype._retrieveMany,
  update = _Connection$prototype.update,
  _updateSingle = _Connection$prototype._updateSingle,
  _updateParallel = _Connection$prototype._updateParallel,
  _updateMany = _Connection$prototype._updateMany,
  upsert = _Connection$prototype.upsert,
  destroy = _Connection$prototype.destroy,
  _destroySingle = _Connection$prototype._destroySingle,
  _destroyParallel = _Connection$prototype._destroyParallel,
  _destroyMany = _Connection$prototype._destroyMany,
  describe = _Connection$prototype.describe,
  describeGlobal = _Connection$prototype.describeGlobal,
  sobject = _Connection$prototype.sobject;
var describeCacheKey = function describeCacheKey(type) {
  return type ? "describe.".concat(type) : 'describe';
};

/**
 * API class for Tooling API call
 */
export var Tooling = /*#__PURE__*/function () {
  /**
   *
   */
  function Tooling(conn) {
    _classCallCheck(this, Tooling);
    /**
     * Execute query by using SOQL
     */
    _defineProperty(this, "query", query);
    /**
     * Query next record set by using query locator
     */
    _defineProperty(this, "queryMore", queryMore);
    _defineProperty(this, "_ensureVersion", _ensureVersion);
    /**
     * Create records
     */
    _defineProperty(this, "create", create);
    _defineProperty(this, "_createSingle", _createSingle);
    _defineProperty(this, "_createParallel", _createParallel);
    _defineProperty(this, "_createMany", _createMany);
    /**
     * Synonym of Tooling#create()
     */
    _defineProperty(this, "insert", create);
    /**
     * Retrieve specified records
     */
    _defineProperty(this, "retrieve", retrieve);
    _defineProperty(this, "_retrieveSingle", _retrieveSingle);
    _defineProperty(this, "_retrieveParallel", _retrieveParallel);
    _defineProperty(this, "_retrieveMany", _retrieveMany);
    /**
     * Update records
     */
    _defineProperty(this, "update", update);
    _defineProperty(this, "_updateSingle", _updateSingle);
    _defineProperty(this, "_updateParallel", _updateParallel);
    _defineProperty(this, "_updateMany", _updateMany);
    /**
     * Upsert records
     */
    _defineProperty(this, "upsert", upsert);
    /**
     * Delete records
     */
    _defineProperty(this, "destroy", destroy);
    _defineProperty(this, "_destroySingle", _destroySingle);
    _defineProperty(this, "_destroyParallel", _destroyParallel);
    _defineProperty(this, "_destroyMany", _destroyMany);
    /**
     * Synonym of Tooling#destroy()
     */
    _defineProperty(this, "delete", destroy);
    /**
     * Synonym of Tooling#destroy()
     */
    _defineProperty(this, "del", destroy);
    _defineProperty(this, "cache", new Cache());
    /**
     * Describe SObject metadata
     */
    _defineProperty(this, "describe", this.cache.createCachedFunction(describe, this, {
      key: describeCacheKey,
      strategy: 'NOCACHE'
    }));
    _defineProperty(this, "describe$", this.cache.createCachedFunction(describe, this, {
      key: describeCacheKey,
      strategy: 'HIT'
    }));
    _defineProperty(this, "describe$$", this.cache.createCachedFunction(describe, this, {
      key: describeCacheKey,
      strategy: 'IMMEDIATE'
    }));
    /**
     * Synonym of Tooling#describe()
     */
    _defineProperty(this, "describeSObject", this.describe);
    _defineProperty(this, "describeSObject$", this.describe$);
    _defineProperty(this, "describeSObject$$", this.describe$$);
    /**
     * Describe global SObjects
     */
    _defineProperty(this, "describeGlobal", this.cache.createCachedFunction(describeGlobal, this, {
      key: 'describeGlobal',
      strategy: 'NOCACHE'
    }));
    _defineProperty(this, "describeGlobal$", this.cache.createCachedFunction(describeGlobal, this, {
      key: 'describeGlobal',
      strategy: 'HIT'
    }));
    _defineProperty(this, "describeGlobal$$", this.cache.createCachedFunction(describeGlobal, this, {
      key: 'describeGlobal',
      strategy: 'IMMEDIATE'
    }));
    /**
     * Get SObject instance
     */
    _defineProperty(this, "sobject", sobject);
    _defineProperty(this, "sobjects", {});
    this._conn = conn;
  }

  /**
   * @private
   */
  return _createClass(Tooling, [{
    key: "version",
    get: function get() {
      return this._conn.version;
    }
  }, {
    key: "_establish",
    value: function _establish() {
      var _this = this;
      this.sobjects = {};
      this.cache.clear();
      this.cache.get('describeGlobal').removeAllListeners('value');
      this.cache.get('describeGlobal').on('value', function (res) {
        if (res.result) {
          var _iterator = _createForOfIteratorHelper(res.result.sobjects),
            _step;
          try {
            for (_iterator.s(); !(_step = _iterator.n()).done;) {
              var type = _step.value.name;
              _this.sobject(type);
            }
          } catch (err) {
            _iterator.e(err);
          } finally {
            _iterator.f();
          }
        }
      });
    }

    /**
     * @private
     */
  }, {
    key: "_baseUrl",
    value: function _baseUrl() {
      return this._conn._baseUrl() + '/tooling';
    }

    /**
     * @private
     */
  }, {
    key: "_supports",
    value: function _supports(feature) {
      return this._conn._supports(feature);
    }

    /**
     *
     */
  }, {
    key: "request",
    value: function request(_request, options) {
      return this._conn.request(_request, options);
    }

    /**
     * Executes Apex code anonymously
     */
  }, {
    key: "executeAnonymous",
    value: function executeAnonymous(body) {
      var url = this._baseUrl() + '/executeAnonymous?anonymousBody=' + encodeURIComponent(body);
      return this.request(url);
    }

    /**
     * Executes Apex tests asynchronously
     */
  }, {
    key: "runTestsAsynchronous",
    value: function runTestsAsynchronous(req) {
      var url = this._baseUrl() + '/runTestsAsynchronous/';
      return this._conn.requestPost(url, req);
    }

    /**
     * Executes Apex tests synchronously
     */
  }, {
    key: "runTestsSynchronous",
    value: function runTestsSynchronous(req) {
      var url = this._baseUrl() + '/runTestsSynchronous/';
      return this._conn.requestPost(url, req);
    }

    /**
     * Retrieves available code completions of the referenced type
     */
  }, {
    key: "completions",
    value: function completions() {
      var type = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'apex';
      var url = this._baseUrl() + '/completions?type=' + encodeURIComponent(type);
      return this.request({
        method: 'GET',
        url: url,
        headers: {
          Accept: 'application/json'
        }
      });
    }
  }]);
}();

/*--------------------------------------------*/
/*
 * Register hook in connection instantiation for dynamically adding this API module features
 */
registerModule('tooling', function (conn) {
  return new Tooling(conn);
});
export default Tooling;
